### PR TITLE
feat(customization): plugin system

### PR DESCRIPTION
**Pull request by romain-gilliotte**
### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)
- [ ] Ensure that Types have been updated according to your changes (if needed)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: docs/agent-customization/plugins/custom.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+Each plugin is nothing more than an `async function` which can perform customizations at either agent level, collection level or both.
+
+```javascript
+export async function removeTimestamps(dataSource, collection, options) {
```
Comment (Scra3) [inline docs/agent-customization/plugins/custom.md:4]: `removeTimestamps` has not the same signature than the previous one.

File: docs/agent-customization/plugins/custom.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+| agent<br>([api-reference](https://forestadmin.github.io/agent-nodejs/classes/_forestadmin_datasource_customizer.DataSourceCustomizer.html))      | An object that allows customizing the whole agent. It has the same interface than the `Agent` you manipulate outside of plugins                                                                         |
+| collection<br>([api-reference](https://forestadmin.github.io/agent-nodejs/classes/_forestadmin_datasource_customizer.CollectionCustomizer.html)) | An object that allows customizing the collection that the plugin was called from (null if the plugin was called on the agent). It is the same object than is passed when you call `customizeCollection` |
+| options                                                                                                                                          | Options which were provided when calling the plugin                                                                                                                                                     |
```
Comment (Scra3) [inline docs/agent-customization/plugins/custom.md:15]: What is the content of options ?

File: docs/agent-customization/plugins/custom.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+Relevant documentation:
+
+- [DataSourceSchema](https://forestadmin.github.io/agent-nodejs/types/_forestadmin_datasource_toolkit.DataSourceSchema.html) (for plugins working on charts)
+- [CollectionSchema](https://forestadmin.github.io/agent-nodejs/types/_forestadmin_datasource_toolkit.CollectionSchema.html) (for all other plugins)
+
+```javascript
+export async function removeTimestamps(agent, collection, options) {
+  const fieldsToRemove = [];
+
+  for (const collection of agent.collections) {
```
Comment (Scra3) [inline docs/agent-customization/plugins/custom.md:32]: the `collection` is declared twice. Rename it please :)

File: packages/agent/src/agent.ts [hunk 4]
```diff
@@ -122,6 +123,17 @@ export default class Agent<S extends TSchema = TSchema> extends FrameworkMounter
     return this;
   }
 
+  /**
+   * Load a plugin, agent-wide
+   * @param plugin instance of the plugin
+   * @param options options which need to be passed to the plugin
```
Comment (Scra3) [inline packages/agent/src/agent.ts:129]: can we have a little example here ?

File: packages/agent/test/agent.test.ts [hunk 5]
```diff
@@ -80,6 +82,13 @@ describe('Agent', () => {
       expect(mockCustomizeCollection).toHaveBeenCalledTimes(1);
     });
 
+    test('use should proxy the call', async () => {
```
Comment (Scra3) [inline packages/agent/test/agent.test.ts:85]: What do you test ? 
It seems unnecessary, you test that a mock works.

Comment (Scra3) [reply packages/agent/test/agent.test.ts:85]: Maybe we can test that a plugin is really executed after the start ?

File: packages/datasource-customizer/test/datasource-customizer.test.ts [hunk 6]
```diff
@@ -17,13 +16,55 @@ beforeEach(() => {
   mockWriteFile.mockReset();
 });
 
-describe('Builder > Agent', () => {
+describe('DataSourceCustomizer', () => {
+  const logger = () => {};
+
+  it('schema should return a schema', async () => {
```
Comment (Scra3) [inline packages/datasource-customizer/test/datasource-customizer.test.ts:22]: `applyQueuedCustomizations` should apply the schema customizations.

File: packages/datasource-customizer/test/datasource-customizer.test.ts [hunk 7]
```diff
@@ -17,13 +16,55 @@ beforeEach(() => {
...
+    customizer.addDataSource(async () =>
+      factories.dataSource.build({ schema: { charts: ['foo'] } }),
+    );
+
+    expect(customizer.schema).toStrictEqual({ charts: [] });
+    await customizer.applyQueuedCustomizations(logger);
+    expect(customizer.schema).toStrictEqual({ charts: ['foo'] });
+  });
+
+  it('collections should return an array of customizers', async () => {
```
Comment (Scra3) [inline packages/datasource-customizer/test/datasource-customizer.test.ts:33]: applyQueuedCustomizations should apply the collection customizations.

Comment (Scra3) [reply packages/datasource-customizer/test/datasource-customizer.test.ts:33]: We can do a describe of the `applyQueuedCustomizations`method.

Comment (Scra3) [reply packages/datasource-customizer/test/datasource-customizer.test.ts:33]: Also why this method is public ?

File: packages/datasource-customizer/test/datasource-customizer.test.ts [hunk 8]
```diff
@@ -17,13 +16,55 @@ beforeEach(() => {
...
+    );
+
+    expect(customizer.collections).toHaveLength(0);
+    await customizer.applyQueuedCustomizations(logger);
+    expect(customizer.collections).toHaveLength(1);
+    expect(customizer.collections[0]).toBeInstanceOf(CollectionCustomizer);
+  });
+
+  describe('getFactory', () => {
+    it('should return a factory', async () => {
```
Comment (Scra3) [inline packages/datasource-customizer/test/datasource-customizer.test.ts:46]: Should return a data source factory customized.

File: packages/datasource-customizer/src/collection-customizer.ts [hunk 9]
```diff
@@ -1,46 +1,68 @@
...
+import importField from './plugins/import-field';
 
 export default class CollectionCustomizer<
   S extends TSchema = TSchema,
   N extends TCollectionName<S> = TCollectionName<S>,
 > {
-  private readonly name: string;
+  private readonly dataSourceCustomizer: DataSourceCustomizer<S>;
   private readonly stack: DecoratorsStack;
+  private readonly customizations: ((logger: Logger) => Promise<void>)[];
```
Comment (Scra3) [inline packages/datasource-customizer/src/collection-customizer.ts:31]: image

File: packages/agent/test/agent.test.ts (hunk 5 above)
Comment (romain-gilliotte) [reply packages/agent/test/agent.test.ts:85]: The agent class is a [proxy](https://refactoring.guru/design-patterns/proxy) of the datasource customizer.

I just test that the underlying implementation is getting called
The real work is done in the datasource-customizer package, and tested there

**Pull request merged**